### ISSUE TITLE
Hover may not work for certain type stub import patterns

### ISSUE THREAD

**Issue by ethereon** [microsoft/pyright#8725]
Consider the following toy typestub module:

`strawberry/__init__.pyi`:
```python
from .jam import buy # Variant 1
# from .jam import * # Variant 2
```

`strawberry/jam.pyi` :
```python
def buy(amount: int) -> str:
    """Buy some delicious strawberry jam"""
    ...
```

Pyright seems to provide different hover results depending on whether `Variant 1` (explicit import by name) or `Variant 2` (wildcard import) is used in `__init__.pyi`. 

For instance, requesting hover results at `buy` in the usage `from strawberry import buy`:

- `Variant 1` fails to resolve:
```json
 {
    "contents": {
        "kind": "markdown",
        "value": "```python\n(import) buy: Unknown\n```"
    },
    "range": {
        "start": {
            "line": 0,
            "character": 23
        },
        "end": {
            "line": 0,
            "character": 26
        }
    }
}
  ```
  
  - `Variant 2` works correctly:
  ```json
 {
    "contents": {
        "kind": "markdown",
        "value": "```python\n(function) def buy(amount: int) -> str\n```\n---\nBuy some delicious strawberry jam"
    },
    "range": {
        "start": {
            "line": 0,
            "character": 23
        },
        "end": {
            "line": 0,
            "character": 26
        }
    }
}
  ```

This only seems to occur with `.pyi` files. For regular Python modules, both variants seem to work correctly.

Tested with `Pyright language server 1.1.375` / recent main (https://github.com/microsoft/pyright/commit/de0386d690a85d844e721032fc517c4403a98984)

**Comment by erictraut** [comment]
Yes, this is by design and in conformance with the Python typing spec. The rules for `.pyi` files are defined [here](https://typing.readthedocs.io/en/latest/spec/distributing.html#stub-files). If you use the `from .jam import buy`, a type checker assumes that `buy` was imported for local use in `__init__.pyi` and is not intended to be re-exported. If you want to re-export it, you need to use a redundant import form (`from .jam import buy as buy`).
